### PR TITLE
Prepare namespaces for AST refactoring

### PR DESCRIPTION
**Pull request by Stagno**
## Technical Description

_This PR is a preparatory step for the upcoming AST refactoring. It touches most of the code files, while the AST refactoring itself won't. Therefore this should be a PR of its own._

Moves AST classes to a new AST module. SIR/AST*.h and IIR/AST*.h headers serve as a link to those classes. Adding aliases from `dawn::sir` and `dawn::iir` namespaces to `dawn::ast` namespace for AST classes such as `AST`, `Stmt`, `BlockStmt`, `Expr`, `ASTVisitor` and so on. Prefixing with the correct namespace all usages of such classes in all code files where they are needed.

#### Example

`dawn::BlockStmt` becomes `dawn::sir::BlockStmt` for code that deals with sir's AST and `dawn::iir::BlockStmt` for code that deals with iir's AST. For now, both `dawn::sir::BlockStmt` and `dawn::iir::BlockStmt` are aliases to `dawn::ast::BlockStmt`.

#### Testing
Tests run from gtclang's PR https://github.com/MeteoSwiss-APN/gtclang/pull/151

### CODE REVIEW

**Comment by Stagno** [comment]
launch jenkins

File: src/dawn/IIR/AST.h [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+//===------------------------------------------------------------------------------------------===//
+
+#ifndef DAWN_IIR_AST_H
+#define DAWN_IIR_AST_H
+
+#include "dawn/AST/AST.h"
+
+namespace dawn {
+namespace iir {
+using AST = ast::AST;
```
Comment (twicki) [inline src/dawn/IIR/AST.h]: should we mark this as temporary?

File: src/dawn/IIR/ASTExpr.h [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+//===------------------------------------------------------------------------------------------===//
+
+#ifndef DAWN_IIR_ASTEXPR_H
+#define DAWN_IIR_ASTEXPR_H
+
+#include "dawn/AST/ASTExpr.h"
+
+namespace dawn {
+namespace iir {
+using Expr = ast::Expr;
```
Comment (twicki) [inline src/dawn/IIR/ASTExpr.h]: and this also?

File: src/dawn/IIR/AST.h [hunk 3]
```diff
@@ -0,0 +1,26 @@
+//===--------------------------------------------------------------------------------*- C++ -*-===//
```
Comment (twicki) [inline src/dawn/IIR/AST.h]: What is the point oif the IIR-AST header?

File: src/dawn/IIR/ASTStringifier.h [hunk 4]
```diff
@@ -0,0 +1,35 @@
+//===--------------------------------------------------------------------------------*- C++ -*-===//
```
Comment (twicki) [inline src/dawn/IIR/ASTStringifier.h]: do we need multiple stringifiers? Wouldn't the one on the AST-Level be enough?

File: src/dawn/IIR/AccessToNameMapper.h [hunk 5]
```diff
@@ -15,7 +15,8 @@
 #ifndef DAWN_IIR_ACCESSTONAMEMAPPER_H
 #define DAWN_IIR_ACCESSTONAMEMAPPER_H
 
-#include "dawn/SIR/ASTVisitor.h"
+#include "dawn/IIR/ASTFwd.h"
```
Comment (twicki) [inline src/dawn/IIR/AccessToNameMapper.h]: why do we now need both includes?

File: src/dawn/IIR/StatementAccessesPair.cpp [hunk 6]
```diff
@@ -38,7 +38,7 @@ static std::string toStringImpl(const StatementAccessesPair* pair,
 
   ss << initialIndentStr << "[\n" << curIndentStr << "Statement:\n";
   ss << "\e[1m"
-     << ASTStringifer::toString(pair->getStatement()->ASTStmt, curIndent + DAWN_PRINT_INDENT)
+     << ASTStringifier::toString(pair->getStatement()->ASTStmt, curIndent + DAWN_PRINT_INDENT)
```
Comment (twicki) [inline src/dawn/IIR/StatementAccessesPair.cpp]: good catch

File: src/dawn/Optimizer/PassInlining.cpp [hunk 7]
```diff
@@ -285,7 +285,7 @@ class Inliner : public ASTVisitor {
         std::iter_swap(newStmtAccessesPairs_.begin() + stmtIdxOfFunc,
                        std::prev(newStmtAccessesPairs_.end()));
 
-        replaceOldExprWithNewExprInStmt(
+        iir::replaceOldExprWithNewExprInStmt(
```
Comment (twicki) [inline src/dawn/Optimizer/PassInlining.cpp]: this should not be specific but just take both of them (since it takes base ptr's. Otherwise we're duplicating code

File: src/dawn/Optimizer/PassTemporaryToStencilFunction.cpp [hunk 8]
```diff
@@ -161,9 +163,10 @@ class TmpAssignment : public ASTVisitorPostOrder, public NonCopyable {
   std::vector<int> accessIDs_; // accessIDs of the accesses that form the tmp = ... expression, that
                                // will become arguments of the stencil fn
 
-  std::shared_ptr<FieldAccessExpr> tmpFieldAccessExpr_ = nullptr; // the field access expr of the
-                                                                  // temporary that is captured and
-                                                                  // being replaced by stencil fn
+  std::shared_ptr<iir::FieldAccessExpr> tmpFieldAccessExpr_ =
+      nullptr;                             // the field access expr of the
```
Comment (twicki) [inline src/dawn/Optimizer/PassTemporaryToStencilFunction.cpp]: something about this formatting strikes me as iffy.

File: src/dawn/Optimizer/Renaming.h [hunk 9]
```diff
@@ -15,6 +15,7 @@
 #ifndef DAWN_OPTIMIZER_RENAMING_H
 #define DAWN_OPTIMIZER_RENAMING_H
 
+#include "dawn/IIR/ASTFwd.h"
```
Comment (twicki) [inline src/dawn/Optimizer/Renaming.h]: why is this here?

File: src/dawn/SIR/ASTExpr.h [hunk 10]
```diff
@@ -15,548 +15,23 @@
...
-  BuiltinTypeID& getBuiltinType() { return builtinType_; }
-
-  virtual std::shared_ptr<Expr> clone() const override;
-  virtual bool equals(const Expr* other) const override;
-  static bool classof(const Expr* expr) { return expr->getKind() == EK_LiteralAccessExpr; }
-  ACCEPTVISITOR(Expr, LiteralAccessExpr)
-};
-
+namespace sir {
+using Expr = ast::Expr;
```
Comment (twicki) [inline src/dawn/SIR/ASTExpr.h]: mark as tmp?

File: src/dawn/SIR/ASTStringifier.h [hunk 11]
```diff
@@ -15,36 +15,22 @@
...
-
-/// @name Stream operators
-/// @ingroup sir
-/// @{
-extern std::ostream& operator<<(std::ostream& os, const AST& ast);
-extern std::ostream& operator<<(std::ostream& os, const std::shared_ptr<Stmt>& expr);
-extern std::ostream& operator<<(std::ostream& os, const std::shared_ptr<Expr>& stmt);
-/// @}
-
+namespace sir {
```
Comment (twicki) [inline src/dawn/SIR/ASTStringifier.h]: again, I think this should be AST level

File: src/dawn/SIR/Statement.h [hunk 12]
```diff
@@ -15,6 +15,7 @@
 #ifndef DAWN_SIR_STATEMENT_H
 #define DAWN_SIR_STATEMENT_H
 
+#include "dawn/SIR/ASTStmt.h"
```
Comment (twicki) [inline src/dawn/SIR/Statement.h]: should be a fwd-declaration

File: src/dawn/Serialization/ASTSerializer.h [hunk 13]
```diff
@@ -23,6 +24,7 @@
 #include <stack>
 
 using namespace dawn;
+using namespace ast;
```
Comment (twicki) [inline src/dawn/Serialization/ASTSerializer.h]: using namespace in a header is sketchy

File: src/dawn/IIR/AST.h (hunk 1 above)
Comment (Stagno) [reply src/dawn/IIR/AST.h]: yes

File: src/dawn/IIR/ASTExpr.h (hunk 2 above)
Comment (Stagno) [reply src/dawn/IIR/ASTExpr.h]: yes

File: src/dawn/IIR/AST.h (hunk 3 above)
Comment (Stagno) [reply src/dawn/IIR/AST.h]: IIR/, SIR/ and AST/ header files of the kind `AST*.h` all have the same mirrored structure.
If some piece of code that works with IIR was including, before the PR, "SIR/AST.h" now it includes "IIR/AST.h". Always including ASTFwd.h may be bad for compilation time, as, to provide aliases, ASTFwd.h includes a lot of other headers.

File: src/dawn/IIR/ASTStringifier.h (hunk 4 above)
Comment (Stagno) [reply src/dawn/IIR/ASTStringifier.h]: There is only one AST stringifier. In this file there is only the alias.

File: src/dawn/IIR/AccessToNameMapper.h (hunk 5 above)
Comment (Stagno) [reply src/dawn/IIR/AccessToNameMapper.h]: IIR/ASTFwd.h to get forward declarations of Stmt, BlockStmt, Expr, ... while IIR/ASTVisitor.h to get the full declaration of the base ASTVisitor.

File: src/dawn/IIR/StatementAccessesPair.cpp (hunk 6 above)
Comment (Stagno) [reply src/dawn/IIR/StatementAccessesPair.cpp]: thx!

File: src/dawn/Optimizer/PassInlining.cpp (hunk 7 above)
Comment (Stagno) [reply src/dawn/Optimizer/PassInlining.cpp]: ptr's to iir::Stmt and ptr's to sir::Stmt cannot be cast one to each other, since they will be different specializations of the same class. But `replaceOldExprWithNewExprInStmt()` is going to be templated in the same way to avoid code duplication, and have its iir:: and sir:: specializations define on which Stmt specialization to operate on.

File: src/dawn/Optimizer/PassTemporaryToStencilFunction.cpp (hunk 8 above)
Comment (Stagno) [reply src/dawn/Optimizer/PassTemporaryToStencilFunction.cpp]: ClangFormat did this ¯\_(ツ)_/¯

File: src/dawn/Optimizer/Renaming.h (hunk 9 above)
Comment (Stagno) [reply src/dawn/Optimizer/Renaming.h]: need fwd definition of iir::Expr

File: src/dawn/SIR/ASTExpr.h (hunk 10 above)
Comment (Stagno) [reply src/dawn/SIR/ASTExpr.h]: yes

File: src/dawn/SIR/ASTStringifier.h (hunk 11 above)
Comment (Stagno) [reply src/dawn/SIR/ASTStringifier.h]: https://github.com/MeteoSwiss-APN/dawn/pull/246#discussion_r317118016

File: src/dawn/SIR/Statement.h (hunk 12 above)
Comment (Stagno) [reply src/dawn/SIR/Statement.h]: need the definition due to https://github.com/MeteoSwiss-APN/dawn/blob/931abab96bcd90be2850782045135d5ab195d73b/src/dawn/SIR/Statement.h#L48

File: src/dawn/Serialization/ASTSerializer.h (hunk 13 above)
Comment (Stagno) [reply src/dawn/Serialization/ASTSerializer.h]: yes, definitely. will fix asap.

File: src/dawn/IIR/ASTStringifier.h (hunk 4 above)
Comment (twicki) [reply src/dawn/IIR/ASTStringifier.h]: yes, but since there is no specialization there should not be aliasing. Just use the `ast::ASTStringifier`

File: src/dawn/SIR/Statement.h (hunk 12 above)
Comment (twicki) [reply src/dawn/SIR/Statement.h]: statement will be gone shortly anyways. This is fine
